Add a debug message, which appears when event loop processing in Mixxx application takes very long

JoergAtGithub opened the pull request.

JoergAtGithub committed:
Add a debug message, which appears when event in Mixxx application took longer than 20ms

daschuer commented on src/mixxxapplication.cpp:
```diff
@@ -61,6 +61,8 @@ class QMouseEventEditable : public QMouseEvent {
 #endif
 };
 
+constexpr mixxx::Duration kEventNotifyExecTimeWarningThreshold = mixxx::Duration::fromMillis(20);
```
What is the reasoning behind this 20 ms? Can you add a comment? A Waveform cycle is 1/60 s = 16.6 ms

JoergAtGithub replied:
I adjusted the threshold to 5ms. This prints more messages on my system, but I think all of them are indicating visible performance issues.
After reducing the limit I see many warnings about slow processing of paint events in KnobComposed widgets. But these disapear, if I apply the fix from #12904 . So I think these are real and no false positives.

JoergAtGithub replied:
I adjusted this again, to 10ms, as there are still many events in the 5ms range after fixing the pixmap cache.

daschuer commented on src/mixxxapplication.cpp:
```diff
@@ -168,7 +170,23 @@ bool MixxxApplication::notify(QObject* target, QEvent* event) {
+    bool ret = QApplication::notify(target, event);
+
+    if (time.elapsed() > kEventNotifyExecTimeWarningThreshold) {
+        qDebug() << "Processing event type"
```
This debug message makes this issue worse because of the extra time it consumes. 
Is it useful to have this active on the user machine, or should this be a compile option for debug builds only?

daschuer replied:
Maybe with CmdlineArgs::Instance().getDeveloper() ?

JoergAtGithub replied:
Done

JoergAtGithub committed:
Added thread info

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.4' into SlowEventLoopNotify

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.5' into SlowEventLoopNotify

JoergAtGithub committed:
Only start the event loop performance timer if we are in developer mode

JoergAtGithub committed:
Adjusted the warning threshold to 5ms and added a comment, which explains this

JoergAtGithub commented:
@daschuer A friendly Ping

ronso0 commented on src/mixxxapplication.cpp:
```diff
@@ -181,10 +182,15 @@ bool MixxxApplication::notify(QObject* target, QEvent* event) {
 
     PerformanceTimer time;
-    time.start();
+    bool isDeveloper = CmdlineArgs::Instance().getDeveloper();
```
I'm curious: isquerying a bool for each event as fast as using a (const) bool member for this constant property?

JoergAtGithub replied:
I cache this in a member variable now, which is initialized in the constructor.

Swiftb0y replied:
> I'm curious: isquerying a bool for each event as fast as using a (const) bool member for this constant property?

If the entire thing is completely inlined, maybe. but caching makes sense here.

JoergAtGithub committed:
Introduce m_isDeveloper member variable in MixxxApplication to cache
developer mode status, reducing redundant calls to
CmdlineArgs::Instance().getDeveloper().

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.5' into SlowEventLoopNotify

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.5' into SlowEventLoopNotify

JoergAtGithub committed:
Adjusted warning threshold again to 10ms

daschuer reviewed (approved):
LGTM, Thank you.

daschuer commented:
Do you want to re-base it on 2.5 before merge? I have no strong demands.

JoergAtGithub commented:
It should be on 2.5 already

daschuer reviewed (approved):
LGTM, Thank you.
